### PR TITLE
Add NVIDIA transcoding support.

### PR DESCRIPTION
**Pull request by j0sh**
**What does this pull request do? Explain your changes. (required)**
Adds nvidia transcoding support.

**Specific updates (required)**
- Updates LPMS to add nvidia transcoding (see https://github.com/livepeer/lpms/pull/114)
- Adds a new `NvidiaTranscoder` type implementing the `Transcoder` interface
- Adds a `-nvidia` command line flag taking a list of GPU device IDs to use for transcoding.
- Updates documentation with a new `doc/gpu.md` page with some usage details for the system.

**How did you test each of these updates (required)**
Unit testing, manual testing on both OSX (software only) and on GCP with four K80s and a P100.

CircleCI is green; AppVeyor is red (ignore).

**Does this pull request close any open issues?**

**Checklist:**

- [x] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

File: doc/gpu.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+Currently the following limitations are observed:
+
+* **Device validity** Ensure valid devices are selected when starting up the node. Currently there is no start-up check to ensure device validity.
+
+* **YUV 4:2:0 input format** The pixel format of the source video must be in YUV 4:2:0 format (planar or
+interleaved). Anything else will return an error.
+
+* **CUDA Availability** If running the Livepeer binary, the CUDA shared libraries are expected to be installed in `/usr/local/cuda`. If the CUDA location differs on your machine, run the node with `LD_LIBRARY_PATH=</path/to/cuda>` environment variable. So far, Livepeer has only been tested with CUDA version 10.0.130.
+
+* **Driver Limits** Retail GPU cards may limit the number of concurrent transcode sessions allowed on the system. This is a soft limit and can be bypassed with an unofficial software patch to the nvidia drivers.
```
Comment (j0sh) [inline doc/gpu.md]: I'm not sure if we should even be mentioning this

File: core/transcoder.go [hunk 2]
```diff
@@ -47,6 +51,64 @@ func NewLocalTranscoder(workDir string) Transcoder {
...
+			glog.Error("Cannot read transcoded output for ", oname)
+			return [][]byte{}, err
+		}
+		out[i] = o
+		os.Remove(oname)
+	}
+	return out, nil
+}
+
+func NewNvidiaTranscoder(inp string, workDir string) Transcoder {
```
Comment (darkdarkdragon) [inline core/transcoder.go]: From name `inp` it's impossible to guess what it should contain. Probably better be something like `devicesList`?

Comment (j0sh) [reply core/transcoder.go]: Fixed

**Comment by darkdarkdragon** [comment]
LGTM
(there is some glitch, I don't see 'Approve' button anymore)

File: core/transcoder.go [hunk 3]
```diff
@@ -48,6 +52,64 @@ func NewLocalTranscoder(workDir string) Transcoder {
 	return &LocalTranscoder{workDir: workDir}
 }
 
+type NvidiaTranscoder struct {
+	workDir string
+	devices []string
+
+	// The following fields need to be protected by the mutex `mu`
+	mu     *sync.Mutex
+	curDev int
```
Comment (angyangie) [inline core/transcoder.go]: What does `curDev` stand for? Perhaps we can use a more descriptive name here

**Review by angyangie — APPROVED** [review]

File: core/transcoder.go (hunk 3 above)
Comment (j0sh) [reply core/transcoder.go]: current device

Comment (j0sh) [reply core/transcoder.go]: renamed to `devIdx` since it's actually the index within the device list . Also added a comment explaining that.

**Comment by iameli** [comment]
I'm unopinionated as to the Go implementation of course, but I've tested this on two different hardware configurations without incident. :shipit:

**Pull request merged**